treeherder/webapp/api/performance_data.py

```diff
@@ -267,24 +274,78 @@ class AlertSummaryPagination(pagination.PageNumberPagination):
     page_size = 10
 
 
+class PerformanceAlertSummaryFilter(django_filters.FilterSet):
+    id = django_filters.NumberFilter(field_name='id')
+    status = django_filters.NumberFilter(field_name='status')
+    framework = django_filters.NumberFilter(field_name='framework')
+    repository = django_filters.NumberFilter(field_name='repository')
+    alerts__series_signature = django_filters.NumberFilter(field_name='alerts__series_signature')
+    filter_text = django_filters.CharFilter(method='_filter_text')
+    hide_improvements = django_filters.BooleanFilter(method='_hide_improvements')
+    hide_downstream = django_filters.BooleanFilter(method='_hide_related_and_invalid')
+    with_assignee = django_filters.CharFilter(method='_with_assignee')
+
+    def _filter_text(self, queryset, name, value):
+        sep = Value(' ')
+        words = value.split(' ')
+
+        contains_all_words = [Q(full_name__contains=word) for word in words]
+
+        # Django's distinct(*fields) isn't supported for MySQL
+        # https://code.djangoproject.com/ticket/17974
+        filtered_summaries = (queryset.annotate(
+            full_name=Concat(
+                'alerts__series_signature__suite', sep,
+                'alerts__series_signature__test', sep,
+                'alerts__series_signature__platform__platform', sep,
+                'alerts__series_signature__extra_options', sep,
+                'bug_number', sep,
+                'push__revision',
+                output_field=CharField()))
+                                      .filter(*contains_all_words)
+                                      .values('id')
+                                      .distinct())
+
+        return queryset.filter(id__in=Subquery(filtered_summaries))
+
+    def _hide_improvements(self, queryset, name, value):
+        return (queryset
+                .annotate(total_regressions=Count('alerts__is_regression'))
+                .filter(alerts__is_regression=True, total_regressions__gte=1))
+
+    def _hide_related_and_invalid(self, queryset, name, value):
```

Comment:
Equivalent raw SQL query, when `Hide downstream / reassigned to / invalid` is checked:
```sql
SELECT `performance_alert_summary`.`id`,
       `performance_alert_summary`.`repository_id`,
       `performance_alert_summary`.`framework_id`,
       `performance_alert_summary`.`prev_push_id`,
       `performance_alert_summary`.`push_id`,
       `performance_alert_summary`.`manually_created`,
       `performance_alert_summary`.`notes`,
       `performance_alert_summary`.`assignee_id`,
       `performance_alert_summary`.`created`,
       `performance_alert_summary`.`first_triaged`,
       `performance_alert_summary`.`last_updated`,
       `performance_alert_summary`.`status`,
       `performance_alert_summary`.`bug_number`,
       `performance_alert_summary`.`bug_updated`,
       `performance_alert_summary`.`issue_tracker_id`,
       `repository`.`id`,
       `repository`.`repository_group_id`,
       `repository`.`name`,
       `repository`.`dvcs_type`,
       `repository`.`url`,
       `repository`.`branch`,
       `repository`.`codebase`,
       `repository`.`description`,
       `repository`.`active_status`,
       `repository`.`performance_alerts_enabled`,
       `repository`.`expire_performance_data`,
       `repository`.`is_try_repo`,
       `repository`.`tc_root_url`,
       `push`.`id`,
       `push`.`repository_id`,
       `push`.`revision`,
       `push`.`author`,
       `push`.`time`
FROM `performance_alert_summary`
         INNER JOIN `repository` ON (`performance_alert_summary`.`repository_id` = `repository`.`id`)
         INNER JOIN `push` ON (`performance_alert_summary`.`push_id` = `push`.`id`)
WHERE (`repository`.`active_status` = 'active' AND `performance_alert_summary`.`status` = 0 AND
       `performance_alert_summary`.`framework_id` = 1 AND NOT (`performance_alert_summary`.`status` IN (1, 2, 3)))
```